"Error while generating citation style"

### JabRef version

Latest development branch build (please note build date below)

### Operating system

GNU / Linux

### Details on version and operating system

Linux Mint 21.2 Victoria base: Ubuntu 22.04 jammy

### Checked with the latest development build (copy version output from About dialog)

- [X] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

JabRef 5.12--2023-11-28--fe07c12
Linux 6.2.0-37-generic amd64 
Java 21.0.1 
JavaFX 21.0.1+6

1. The Preview tab has stopped working.

![image](https://github.com/JabRef/jabref/assets/762246/07a6d0f6-d470-41c7-ba39-35877f5b56f4)

2. Going into preferences/entry preview shows that system has broken too:

![image](https://github.com/JabRef/jabref/assets/762246/d3a9d439-77b7-4b85-8c5f-109faa2d7d62)



### Appendix

...
<details>

<summary>Log File</summary>

```
Paste an excerpt of your log file here
```

</details>

JabRef 5.12-PullRequest10682.168--2023-12-04--cd3288f
Linux 6.2.0-37-generic amd64 
Java 21.0.1 
JavaFX 21.0.1+6

I've taken that custom build, and I get this error message in the preferences/entry preview/current preview preview window for ACM Materials Journal, the very first citation style.

This error appears for _all_ citation styles with the exception of the default "customized preview style", which displays normally.


```
Error while generating citation style java.lang.NoClassDefFoundError: Could not initialize class de.undercouch.citeproc.csl.internal.SCitationLayout BibTeX (internal): @article{Smith2016, abstract = {This entry describes a test scenario which may be useful in JabRef. By providing a test entry it is possible to see how certain things will look in this graphical BIB-file mananger.}, address = {Trondheim}, author = {Smith, Bill and Jones, Bob and Williams, Jeff}, comment = {Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint obcaecat cupiditat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.}, doi = {10.1001/bla.blubb}, editor = {Taylor, Phil}, file = {:testentry.pdf:PDF}, issue = {7}, journal = {BibTeX Journal}, keywords = {KeyWord1; KeyWord2; KeyWord3; Keyword4}, month = {July}, number = {3}, pages = {45--67}, publisher = {JabRef Publishing}, title = {Title of the test entry}, url = {https://github.com/JabRef}, volume = {34}, year = {2016}, _jabref_shared = {sharedId: -1, version: 1} } Stack Trace: java.lang.Exception: java.lang.NoClassDefFoundError: Could not initialize class de.undercouch.citeproc.csl.internal.SCitationLayout at org.jabref@5.12.30168/org.jabref.gui.util.DefaultTaskExecutor.convertToException(Unknown Source) at org.jabref@5.12.30168/org.jabref.gui.util.DefaultTaskExecutor.lambda$getJavaFXTask$5(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/javafx.event.Event.fireEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/javafx.concurrent.EventHelper.fireEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/javafx.concurrent.Task.fireEvent(Unknown Source) at org.jabref.merged.module@5.12.30168/javafx.concurrent.Task.setState(Unknown Source) at org.jabref.merged.module@5.12.30168/javafx.concurrent.Task$TaskCallable.lambda$call$2(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source) at java.base/java.security.AccessController.doPrivileged(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source) at org.jabref.merged.module@5.12.30168/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method) at org.jabref.merged.module@5.12.30168/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$10(Unknown Source) at java.base/java.lang.Thread.run(Unknown Source) Caused by: java.lang.NoClassDefFoundError: Could not initialize class de.undercouch.citeproc.csl.internal.SCitationLayout at org.jabref.merged.module@5.12.30168/de.undercouch.citeproc.csl.internal.SCitation.(Unknown Source) at org.jabref.merged.module@5.12.30168/de.undercouch.citeproc.csl.internal.SStyle.(Unknown Source) at org.jabref.merged.module@5.12.30168/de.undercouch.citeproc.CSL.loadStyle(Unknown Source) at org.jabref.merged.module@5.12.30168/de.undercouch.citeproc.CSL.(Unknown Source) at org.jabref@5.12.30168/org.jabref.logic.citationstyle.CSLAdapter.initialize(Unknown Source) at org.jabref@5.12.30168/org.jabref.logic.citationstyle.CSLAdapter.makeBibliography(Unknown Source) at org.jabref@5.12.30168/org.jabref.logic.citationstyle.CitationStyleGenerator.generateCitations(Unknown Source) at org.jabref@5.12.30168/org.jabref.logic.citationstyle.CitationStyleGenerator.generateCitation(Unknown Source) at org.jabref@5.12.30168/org.jabref.logic.citationstyle.CitationStylePreviewLayout.generatePreview(Unknown Source) at org.jabref@5.12.30168/org.jabref.gui.preview.PreviewViewer.lambda$update$4(Unknown Source) at org.jabref@5.12.30168/org.jabref.gui.util.BackgroundTask$1.call(Unknown Source) at org.jabref@5.12.30168/org.jabref.gui.util.DefaultTaskExecutor$1.call(Unknown Source) at org.jabref.merged.module@5.12.30168/javafx.concurrent.Task$TaskCallable.call(Unknown Source) at java.base/java.util.concurrent.FutureTask.run(Unknown Source) at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) at java.base/java.util.concurrent.FutureTask.run(Unknown Source) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ... 1 more Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.IllegalAccessError: class de.undercouch.citeproc.csl.internal.SCitationLayout (in module org.jabref.merged.module) cannot access class org.apache.commons.lang3.CharSet (in module org.apache.commons.lang3) because module org.jabref.merged.module does not read module org.apache.commons.lang3 [in thread "pool-1-thread-2"] at org.jabref.merged.module@5.12.30168/de.undercouch.citeproc.csl.internal.SCitationLayout.(Unknown Source) ... 19 more
```

--- a/build.gradle
+++ b/build.gradle
@@ -592,6 +592,7 @@ jlink {
         requires 'com.google.gson'
         requires 'org.slf4j'
         requires 'jakarta.xml.bind'
+        requires 'org.apache.commons.lang3'
         uses 'org.mariadb.jdbc.credential.CredentialPlugin'
         uses 'org.mariadb.jdbc.authentication.AuthenticationPlugin'
         uses 'org.mariadb.jdbc.tls.TlsSocketPlugin'

--- a/src/main/java/org/jabref/gui/preview/PreviewViewer.java
+++ b/src/main/java/org/jabref/gui/preview/PreviewViewer.java
@@ -1,6 +1,8 @@
 package org.jabref.gui.preview;
 
 import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.net.MalformedURLException;
 import java.util.Objects;
 import java.util.Optional;
@@ -257,13 +259,22 @@ public class PreviewViewer extends ScrollPane implements InvalidationListener {
 
         Number.serialExportNumber = 1; // Set entry number in case that is included in the preview layout.
 
+        final BibEntry theEntry = entry.get();
         BackgroundTask
-                .wrap(() -> layout.generatePreview(entry.get(), database))
+                .wrap(() -> layout.generatePreview(theEntry, database))
                 .onRunning(() -> setPreviewText("<i>" + Localization.lang("Processing %0", Localization.lang("Citation Style")) + ": " + layout.getDisplayName() + " ..." + "</i>"))
                 .onSuccess(this::setPreviewText)
                 .onFailure(exception -> {
                     LOGGER.error("Error while generating citation style", exception);
-                    setPreviewText(Localization.lang("Error while generating citation style"));
+
+                    // Convert stack trace to a string
+                    StringWriter stringWriter = new StringWriter();
+                    PrintWriter printWriter = new PrintWriter(stringWriter);
+                    exception.printStackTrace(printWriter);
+                    String stackTraceString = stringWriter.toString();
+
+                    // Set the preview text with the localized error message and the stack trace
+                    setPreviewText(Localization.lang("Error while generating citation style") + "\n\n" + exception.getLocalizedMessage() + "\n\nBibTeX (internal):\n" + theEntry + "\n\nStack Trace:\n" + stackTraceString);
                 })
                 .executeWith(taskExecutor);
     }

